corehq/apps/smsforms/models.py

```diff
@@ -103,16 +104,31 @@ def close(self):
         if not self.session_is_open:
             return
 
-        self.mark_completed(False)
-
-        if self.submit_partially_completed_forms:
-            submit_unfinished_form(self)
+        try:
+            if self.submit_partially_completed_forms:
+                submit_unfinished_form(self)
+        finally:
+            # this needs to be called after the submission, but regardless of whether it succeeded
+            # thus the try/finally
+            self.mark_completed(False)
```

Comment:
Having this in a `finally` block looks like a change that goes beyond just collecting metrics. I'm curious how you decided it was OK to include it in this PR?

Reply:
Can you articulate how? I think this is okay and I'll describe why, but I want to make sure I'm not missing the red flag you're seeing.

Basically, to be able to emit the metric with details about whether the form was sent upon completion or not, I needed to switch the order of these lines from (mark_completed, submit_unfinished_form) to (submit_unfinished_form, mark_completed). So I set about thinking of whether there was a safe way to do that. I split it up into two cases: (1) an error is raised inside the `if self.submit_partially_completed_forms: submit_unfinished_form(self)` block, and (2) no error is raised there. If no error is raised, then both will execute but just in reverse order. Since `submit_partially_completed_forms` and `submit_unfinished_form` don't rely on of the properties of `session` set in `mark_completed` and vice versa (except for the dependency that's motivating the switch), after both are run, the effect will be the same as it was in the previous ordering. If there _is_ an error, then I would want to make sure `mark_completed` is run anyway, since in the original order, it had already been run, so I wrapped it in a `try` and put `mark_completed` in a `finally` block. In both situations, then, I concluded that the total result would be the same in the new code as in the old code, except that I would be able to emit this new piece of information.

Reply:
It's hard for me to evaluate if these three lines

https://github.com/dimagi/commcare-hq/blob/2c0b4505d2c4d3da16a333ac2bc43e676aeff14d/corehq/apps/smsforms/models.py#L116-L118

have any impact on `submit_unfinished_form(self)`. It seems safer to continue to do those things before, and then call `metrics_counter(...)` in the finally block. It's not clear to me why that was not possible.

Reply:
Oh, I see. `mark_completed` is called in a bunch of different places (not just from here), so I could have worked around by emitting in both places and not emitting in mark_complete if a special parameter was passed, but this seemed more straightforward.

I'm fairly confident from searching the code that `submit_unfinished_form` doesn't depend on any of those three values